### PR TITLE
Fixed #26743 -- Fixed UnboundLocalError crash when deserializing m2m fields and value isn't iterable.

### PR DESCRIPTION
**Pull request by bmispelon**
Even though this change affect all serializers, I only added a test for the `json` one. This seems like a somewhat consistent practice: `tests_json.py` has a lot more test methods than `tests_xml.py` for example and not all of them are json-specific.

Adding the test to the base class (the one in `tests.py` that gets executed for all available serialization formats) is complicated because it's not straightforward to generate the broken serialized data that correspond to the active format (`self.serializer_name`).

ticket-26743

### CODE REVIEW

File: django/core/serializers/base.py [hunk 1]
```diff
@@ -283,15 +283,23 @@ def m2m_convert(v):
...
+        except Exception as e:
+            if isinstance(e, ObjectDoesNotExist) and handle_forward_references:
+                return DEFER_FIELD
+            else:
+                raise M2MDeserializationError(e, pk)
         else:
-            raise M2MDeserializationError(e, pk)
+            values.append(converted)
+
+    return values
```
Comment (felixxm) [inline django/core/serializers/base.py]: Some parts of this change are unrelated with a bugfix :thinking: Can we leave only necessary changes?, i.e.
```diff
diff --git a/django/core/serializers/base.py b/django/core/serializers/base.py
index cf8e7d3b86..fac4440b62 100644
--- a/django/core/serializers/base.py
+++ b/django/core/serializers/base.py
@@ -282,9 +282,14 @@ def deserialize_m2m_values(field, field_value, using, handle_forward_references)
         def m2m_convert(v):
             return model._meta.pk.to_python(v)

+    try:
+        pks_iter = iter(field_value)
+    except TypeError as e:
+        raise M2MDeserializationError(e, field_value)
     try:
         values = []
-        for pk in field_value:
+        for pk in pks_iter:
             values.append(m2m_convert(pk))
         return values
     except Exception as e:
```

File: tests/serializers/test_json.py [hunk 2]
```diff
@@ -252,6 +252,24 @@ def test_helpful_error_message_for_many2many_natural2(self):
...
+            'model': 'serializers.M2MData',
+            'fields': {'data': None},
+        }]
+
+        serialized = json.dumps(data_to_serialize)
+        expected = "(serializers.M2MData:pk=1) field_value was 'None'"
+
+        with self.assertRaisesMessage(DeserializationError, expected):
+            for _ in serializers.deserialize('json', serialized, ignore=False):
+                pass
```
Comment (felixxm) [inline tests/serializers/test_json.py]: IMO we can simplify this with `next()`, i.e.
```python
with self.assertRaisesMessage(DeserializationError, expected):
    next(serializers.deserialize('json', serialized, ignore=False))
```

File: django/core/serializers/base.py (hunk 1 above)
Comment (bmispelon) [reply django/core/serializers/base.py]: Sure, I guess I cleaned up a bit more than what was strictly necessary.

File: tests/serializers/test_json.py (hunk 2 above)
Comment (bmispelon) [reply tests/serializers/test_json.py]: Good idea, thanks.

File: django/core/serializers/base.py [hunk 3]
```diff
@@ -282,9 +282,14 @@ def m2m_convert(value):
         def m2m_convert(v):
             return model._meta.pk.to_python(v)
 
+    try:
+        pks_iter = iter(field_value)
+    except TypeError as e:
+        raise M2MDeserializationError(e, field_value)
```
[the remark this answers is not in the document]
Comment (bmispelon) [reply django/core/serializers/base.py]: I'm not familiar with the whole (de)serializer codebase but it seems to me that when we reach this function then `field_value` should really be iterable.

If we're trying to deserialize `None` or anything un-iterable then something has gone wrong, no? Wouldn't it be better to error out rather than try to correct the issue?

[the remark this answers is not in the document]
Comment (felixxm) [reply django/core/serializers/base.py]: Agreed.

**Comment by felixxm** [comment]
@bmispelon Thanks :star2:

**Comment by bmispelon** [comment]
Thanks! :1st_place_medal:
